### ISSUE TITLE
Need help deciding how to proceed with octo.erl

### ISSUE THREAD

**Issue by Minoru** [codingteam/Gestalt#18]
GitHub API provides us with a paginated list of events. It also has rate limiting (5000 requests per hour for authorized users) and doesn't count requests that resulted in `304 Not Modified` toward the limit. Unfortunately, to get `304` you should specify `If-None-Match` or `If-Modified-Since` header, and `octo.erl` doesn't do that. Furthermore, it doesn't even expose the `ETag` and/or `Last-Modified` headers returned with the response, so we just _can't_ implement rate limit handling on our own.

Our options, as I see them, are:
1. Change return types of all the functions in `octo.erl` so that they return `ETag` and `Last-Modified` headers.
   
   **Cons:**
   - that breaks the public interface of the library;
   - all the clients that wish to handle rate limits will have to write their own code to do that, which is A Bad Thing.
2. Turn `octo.erl` into a server that holds a list of latest responses alongside with corresponding `ETag`s.
   
   **Cons:**
   - that **horribly** breaks the public interface of the library. We might as well start a new one!
   - the change is radical and might not make it into the upstream, which pretty much defeats the point of going with existing implementation in the first place.
3. Bend, and pray to the Emperor that we won't hit the limit.
   
   **Cons:**
   - no amount of prayer will erase the fact that we _will_ hit the limit, probably right on the start of the service when it will try to read as much of organization's history as possible. And don't forget it also got to take a look at user's public repos to calculate the statistics over languages used!
4. Abandon `octo.erl` and roll out our own implementation (just as @ForNeVeR originally suggested).
   
   **Pros:**
   - we can do whatever we please;
   - once we have it working, the code can be used to persuade `octo.erl`'s author that our approach is better, resulting in the library rewrite and our migration over to it.
   
   And I can't see no cons here =\

So right now, I'm leaning toward the fourth option, but would like to hear your comments and thoughts on the matter. And if nothing else, this issue will mark a design decision that we might want to re-evaluate later on.

**Comment by ForNeVeR** [comment]
Well, I can see two realistic cases 2 and 4 here. And to make final decision between them, you need to answer a question: what's the difference between them? Are you sure that your finely-crafted library (case 4) will not become just another clone of octo.erl with minor API changes? If not, then maybe simply fork octo.erl, implement the breaking changes you proposed and then contact authors to see if they would like to create new, incompatible version of their library? If they don't want - then we may proceed with our fork.

That argument is invalid, of course, if you see any major API and concept improvements we can make in our library and cannot do that in our octo.erl fork.

Personally I'd suggest to choose balanced variant between 2 and 4 - create our fork of octo.erl and start our library with that.

Also we should compare amounts of time we will need to implement any of your suggestions. Maybe it's not that hard and we can simply roll out our ad-hoc implementation without care about anything.

**Comment by Minoru** [comment]
I'm currently contemplating how to goad `octo.erl` toward morphing as I described in option 2, without rewriting the whole thing. You'll probably see the results shortly, keep an eye on our fork of the library.

**Comment by Minoru** [comment]
Update: it turned out that all I need to do is convert main module into `gen_server` and then rewrite `octo_http_helper:do_request` so that it caches responses and handles conditional queries. I was halfway through the changes, tracking down the bug that prevented the thing from working, when I went on my little hiatus which then turned into a long computer-free weekend — sorry for disappearing like that!

I'll push my changes the moment I has that bug found, which, hopefully, will occur sometime tomorrow.

**Comment by Minoru** [comment]
Time for the final update!
1. I managed to implement all the changes mentioned in the previous update. `octo.erl` now caches responses along with `ETag` and `Last-Modified` headers, and uses those when sending requests. It's completely transparent to the user and didn't require any change to library's interface.
2. I'm now a maintainer of `octo.erl` and have power to push it into any direction I see fit (well, at least I think so.)
3. I recently implemented Organizations API, so it looks like we can finally start working on Gestalt itself.
